src/hotspot/share/opto/gcm.cpp

```diff
@@ -770,23 +813,7 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
       // Add an anti-dep edge, and squeeze 'load' into the highest block.
       assert(store != load->find_exact_control(load->in(0)), "dependence cycle found");
       if (verify) {
-#ifdef ASSERT
-        // We expect an anti-dependence edge from 'load' to 'store', except when
-        // implicit_null_check() has hoisted 'store' above its early block to
-        // perform an implicit null check, and 'load' is placed in the null
-        // block. In this case it is safe to ignore the anti-dependence, as the
-        // null block is only reached if 'store' tries to write to null.
-        Block* store_null_block = NULL;
-        Node* store_null_check = store->find_out_with(Op_MachNullCheck);
-        if (store_null_check != NULL) {
-          Node* if_true = store_null_check->find_out_with(Op_IfTrue);
-          assert(if_true != NULL, "null check without null projection");
-          Node* null_block_region = if_true->find_out_with(Op_Region);
-          assert(null_block_region != NULL, "null check without null region");
-          store_null_block = get_block_for_node(null_block_region);
-        }
-#endif
-        assert(LCA == store_null_block || store->find_edge(load) != -1,
+        assert(store->find_edge(load) != -1 || unrelated_load_in_store_null_block(store, load),
```

Comment:
I'm probably missing something, but the new code doesn't look equivalent to the old code, unless the bug fix is part of the refactoring.  You removed the check against LCA and added a new dominates check.

Reply:
it is refactoring. both issue has same property in terms of the load being used only in null block most probably by uncommon trap. gcm late scheduling schedule it in null block, later implicit null check moves another store which is unrelated to the load in null block* to null check block. we don't rerun insert_anti_dependence for null_block after these move as it is not required, so we never put an anti-dependence edge load to store here. verification phase cribs about this.

as  issues are same, now i check whether the load is in null block of the store.

*I specifically used unre

Reply:
You seem to have replaced LCA == store_null_block with the equivalent of store_null_block->dominates(get_block_for_node(load)).  Please explain.

Reply:
LCA was schedule late location for the load after gcm schedule late and that is null block for the store.I tried improving the fix
saying if null block for the store dominates load block then ignore the case.

Reply:
Can you think of a case where store_null_check != NULL, and the following are not true?
>    LCA == store_null_block
>    get_block_for_node(null_block_region)->dominates(get_block_for_node(load))

It sounds like you expect LCA == store_null_block to always be true.  Do you have an example where the dominates check would not be true?  If both are are preconditions that are always true, then how about making them asserts?

>    assert(null_block_region != NULL, "null check without null region");
>    assert(LCA == store_null_block, "wrong LCA");
>    assert(get_block_for_node(null_block_region)->dominates(get_block_for_node(load)), "does not dominate");
>    return true;

Reply:
if LCA is above null block(ie. null_check block itself or above it) means it has uses in other places . for null_check block anti dependency will be taken care in implicit null logic. above it doesn't actually matter for our considering case i guess.

i kept dominate logic as it is more broad than equating with LCA (that means even if there are multiple basic blocks in null  path the assert holds good)

Reply:
> i kept dominate logic as it is more broad than equating with LCA (that means even if there are multiple basic blocks in null  path the assert holds good)

In that case, could you add a test case with multiple basic blocks in the null path?  It should crash without your dominates improvement.

Reply:
I didn't say there is a case already. I just said it can even take care of that. problem with using LCA is the new fix has raised LCA, may be i should use LCA_Orig in second case. but do you see problem in using dominate in assert?

Reply:
I agree, will retain the earlier fix. will LCA_Orig for the second assert.